### PR TITLE
ambient: Fix destination principal with WDS

### PR DESCRIPTION
**Pull request by keithmattix**
Grab the service account and trust domain from WDS and use that to create the destination principal

### CODE REVIEW

File: source/extensions/common/workload_discovery/api.cc [hunk 1]
```diff
@@ -46,10 +46,22 @@ Istio::Common::WorkloadMetadataObject convert(const istio::workload::Workload& w
...
   }
+  std::string ns = workload.namespace_();
+  std::string trust_domain = workload.trust_domain();
+  // Trust domain may be elided if it's equal to "cluster.local"
+  if (trust_domain.empty()) {
+    trust_domain = "cluster.local";
+  }
+  // The namespace may be elided if it's equal to "default"
+  if (ns.empty()) {
+    ns = "default";
```
Comment (kyessenov) [inline source/extensions/common/workload_discovery/api.cc:57]: This sounds dangerous to imply default.

File: source/extensions/common/workload_discovery/api.cc [hunk 2]
```diff
@@ -46,10 +46,22 @@ Istio::Common::WorkloadMetadataObject convert(const istio::workload::Workload& w
   default:
     break;
   }
+  std::string ns = workload.namespace_();
+  std::string trust_domain = workload.trust_domain();
+  // Trust domain may be elided if it's equal to "cluster.local"
+  if (trust_domain.empty()) {
+    trust_domain = "cluster.local";
```
Comment (kyessenov) [inline source/extensions/common/workload_discovery/api.cc:53]: cluster.local is strongly discouraged - I think it should be explicit.

File: extensions/common/context.cc [hunk 3]
```diff
@@ -293,9 +293,13 @@ namespace {
 bool extractPeerMetadataFromUpstreamMetadata(const std::string& metadata_type,
                                              flatbuffers::FlatBufferBuilder& fbb) {
   std::string endpoint_labels;
+  std::string identity;
```
Comment (kyessenov) [inline extensions/common/context.cc:296]: Undo these changes?

File: source/extensions/filters/http/istio_stats/istio_stats.cc [hunk 4]
```diff
@@ -95,6 +95,11 @@ extractEndpointMetadata(const StreamInfo::StreamInfo& info) {
     const auto& filter_metadata = upstream_host->metadata()->filter_metadata();
     const auto& it = filter_metadata.find("istio");
     if (it != filter_metadata.end()) {
+      std::string identity = "";
```
Comment (kyessenov) [inline source/extensions/filters/http/istio_stats/istio_stats.cc:98]: I'm confused - this is using EDS, not WDS?

File: source/extensions/common/workload_discovery/api.cc [hunk 5]
```diff
@@ -46,10 +46,22 @@ Istio::Common::WorkloadMetadataObject convert(const istio::workload::Workload& w
   default:
     break;
   }
+  std::string ns = workload.namespace_();
```
Comment (kyessenov) [inline source/extensions/common/workload_discovery/api.cc:49]: `const std::string&` to avoid a copy.

File: source/extensions/common/workload_discovery/api.cc (hunk 1 above)
Comment (keithmattix) [reply source/extensions/common/workload_discovery/api.cc:57]: The comments are straight from the WDS proto

File: source/extensions/common/workload_discovery/api.cc (hunk 2 above)
Comment (keithmattix) [reply source/extensions/common/workload_discovery/api.cc:53]: This is also from the proto. In my tests the trust domain was elided as well

File: extensions/common/context.cc (hunk 3 above)
Comment (keithmattix) [reply extensions/common/context.cc:296]: Good catch; this was from an older implementation

File: source/extensions/filters/http/istio_stats/istio_stats.cc (hunk 4 above)
Comment (keithmattix) [reply source/extensions/filters/http/istio_stats/istio_stats.cc:98]: Also an older implementation I forgot to scrub

File: source/extensions/common/workload_discovery/api.cc (hunk 5 above)
Comment (keithmattix) [reply source/extensions/common/workload_discovery/api.cc:49]: I think this is necessary if we do need the defaulting logic?

**Comment by keithmattix** [comment]
Fixes istio/istio##50225

**Comment by keithmattix** [comment]
@kyessenov please see my responses above; if we need to change WDS across waypoint, ztunnel, and istiod, we can probably tackle that later, but I think the PR as-is represents the current state of the API

File: source/extensions/common/workload_discovery/api.cc (hunk 5 above)
Comment (kyessenov) [reply source/extensions/common/workload_discovery/api.cc:49]: I think you can use `string_view` here, and pull out a constant into `constexpr`?

File: source/extensions/common/workload_discovery/api.cc [hunk 6]
```diff
@@ -46,10 +46,22 @@ Istio::Common::WorkloadMetadataObject convert(const istio::workload::Workload& w
...
+  std::string trust_domain = workload.trust_domain();
+  // Trust domain may be elided if it's equal to "cluster.local"
+  if (trust_domain.empty()) {
+    trust_domain = "cluster.local";
+  }
+  // The namespace may be elided if it's equal to "default"
+  if (ns.empty()) {
+    ns = "default";
+  }
+  const auto identity = "spiffe://" + trust_domain + "/ns/" + workload.namespace_() + "/sa/" +
```
Comment (kyessenov) [inline source/extensions/common/workload_discovery/api.cc:59]: Use `absl::StrCat`.

**Review by kyessenov — CHANGES_REQUESTED** [review]
Minor code comments.

File: source/extensions/common/workload_discovery/api.cc (hunk 5 above)
Comment (keithmattix) [reply source/extensions/common/workload_discovery/api.cc:49]: I think that worked thanks!

File: source/extensions/common/workload_discovery/api.cc [hunk 7]
```diff
@@ -46,10 +46,25 @@ Istio::Common::WorkloadMetadataObject convert(const istio::workload::Workload& w
   default:
     break;
   }
+
+  constexpr absl::string_view defaultNs = "default";
```
Comment (kyessenov) [inline source/extensions/common/workload_discovery/api.cc:50]: can you pull them out of and use correct style, e.g. `constexpr absl::string_view DefaultNamespace = ...`

File: source/extensions/common/workload_discovery/api.cc [hunk 8]
```diff
@@ -46,10 +46,25 @@ Istio::Common::WorkloadMetadataObject convert(const istio::workload::Workload& w
...
+  absl::string_view trust_domain = workload.trust_domain();
+  // Trust domain may be elided if it's equal to "cluster.local"
+  if (trust_domain.empty()) {
+    trust_domain = defaultTrustDomain;
+  }
+  // The namespace may be elided if it's equal to "default"
+  if (ns.empty()) {
+    ns = defaultNs;
+  }
+  const auto identity = absl::StrCat("spiffe://", trust_domain, "/ns/" + workload.namespace_(),
```
Comment (kyessenov) [inline source/extensions/common/workload_discovery/api.cc:62]: "/ns/" + <--- please fix

File: test/envoye2e/driver/otel.go [hunk 9]
```diff
@@ -114,6 +114,7 @@ func (v *verify) Run(p *Params) error {
 		return fmt.Errorf("timed out waiting for all metrics to match")
 	}
 }
+
```
Comment (kyessenov) [inline test/envoye2e/driver/otel.go:117]: revert?

File: test/envoye2e/tcp_metadata_exchange/tcp_metadata_exchange_test.go [hunk 10]
```diff
@@ -66,27 +66,30 @@ func TestTCPMetadataExchange(t *testing.T) {
...
 workload_name: productpage-v1
 workload_type: DEPLOYMENT
 canonical_name: productpage-v1
 canonical_revision: version-1
 cluster_id: client-cluster
 uid: //v1/pod/default/productpage
-`}, {
-						Address: "127.0.0.2",
-						Metadata: `
+`,
```
Comment (kyessenov) [inline test/envoye2e/tcp_metadata_exchange/tcp_metadata_exchange_test.go:80]: revert superfluous changes please

File: test/envoye2e/driver/otel.go (hunk 9 above)
Comment (keithmattix) [reply test/envoye2e/driver/otel.go:117]: They were added by make lint; I'll try and revert

File: source/extensions/common/workload_discovery/api.cc (hunk 7 above)
Comment (keithmattix) [reply source/extensions/common/workload_discovery/api.cc:50]: Pull them out to where? I skimmed through the Envoy style guide and didn't see what the proper scoping was. I'll add them to the namespace, but let me know if they need to be elsewhere

**Review by kyessenov — APPROVED** [review]
